diff --git a/engine/logic_monad.ml b/engine/logic_monad.ml
index 1b7f2799e2..11f3621a88 100644
--- a/engine/logic_monad.ml
+++ b/engine/logic_monad.ml
@@ -149,6 +149,14 @@ type ('a, 'b, 'e) list_view_ =
 
 type ('a, 'b, 'e) list_view = ('a, 'e -> 'b, 'e) list_view_
 
+type 'a thunk =
+  | Fin of 'a
+  | TailCall of (unit -> 'a thunk)
+
+let rec trampoline r = match r with
+  | Fin a -> a
+  | TailCall f -> trampoline (f ())
+
 module BackState =
 struct
 
@@ -184,7 +192,7 @@ struct
       [split] is pattern-matching. *)
 
   type ('a, 'i, 'o, 'e) t =
-    { iolist : 'r. 'i -> ('e -> 'r) -> ('a -> 'o -> ('e -> 'r) -> 'r) -> 'r }
+    { iolist : 'r. 'i -> ('e -> 'r) -> ('a -> 'o -> ('e -> 'r) -> 'r thunk) -> 'r thunk }
   (* IMPORTANT: to play well with side-effects, all functions involved in the
      above type must have AT LEAST the same arity as the one implied by their
      type. This is to ensure that applying them to the expected arguments only
@@ -203,7 +211,7 @@ struct
      this property only in this module. *)
 
   let return x =
-    { iolist = fun s nil cons -> cons x s nil }
+    { iolist = fun s nil cons -> TailCall (fun () -> cons x s nil) }
 
   let (>>=) m f =
     { iolist = fun s nil cons ->
@@ -217,10 +225,10 @@ struct
     { iolist = fun s nil cons -> m.iolist s nil (fun x s next -> cons (f x) s next) }
 
   let zero e =
-    { iolist = fun _ nil cons -> nil e }
+    { iolist = fun _ nil cons -> Fin (nil e) }
 
   let plus m1 m2 =
-    { iolist = fun s nil cons -> m1.iolist s (fun e -> (m2 e).iolist s nil cons) cons }
+    { iolist = fun s nil cons -> m1.iolist s (fun e -> trampoline @@ (m2 e).iolist s nil cons) cons }
 
   let ignore m =
     { iolist = fun s nil cons -> m.iolist s nil (fun _ s next -> cons () s next) }
@@ -261,20 +269,20 @@ struct
       and Terminating Monad Transformers" paper.  *)
   type ('a, 'e) reified = { r : ('a, ('a, 'e) reified, 'e) list_view } [@@unboxed]
 
-  let rec reflect0 : type r. _ -> (_ -> r) -> (_ -> _ -> (_ -> r) -> r) -> r =
+  let rec reflect0 : type r. _ -> (_ -> r) -> (_ -> _ -> (_ -> r) -> r thunk) -> r thunk =
     fun m nil cons ->
       match m.r with
-      | Nil e -> nil e
-      | Cons ((x, s), l) -> cons x s (fun e -> reflect0 (l e) nil cons)
+      | Nil e -> Fin (nil e)
+      | Cons ((x, s), l) -> cons x s (fun e -> trampoline @@ reflect0 (l e) nil cons)
 
   let reflect (m : ('a * 'o, 'e) reified) =
     { iolist = fun s0 nil cons -> reflect0 m nil cons }
 
   let split m : (_ list_view, _, _, _) t =
     let rnil e = Nil e in
-    let rcons p s l = Cons ((p, s), (fun e -> {r=l e})) in
+    let rcons p s l = Fin (Cons ((p, s), (fun e -> {r=l e}))) in
     { iolist = fun s nil cons ->
-      begin match m.iolist s rnil rcons with
+      begin match trampoline @@ m.iolist s rnil rcons with
       | Nil e -> cons (Nil e) s nil
       | Cons ((x, s), l) ->
         let l e = reflect (l e) in
@@ -285,9 +293,9 @@ struct
     let rnil e = {r=Nil e} in
     let rcons x s l =
       let p = (x, s) in
-      {r=Cons (p, l)}
+      Fin {r=Cons (p, l)}
     in
-    m.iolist s rnil rcons
+    trampoline @@ m.iolist s rnil rcons
 
   let repr x = x.r
 end
@@ -396,8 +404,8 @@ struct
     let rnil e = {r=Nil e} in
     let rcons x s l =
       let p = (x, s.sstate, s.wstate, s.ustate) in
-      {r=Cons (p, l)}
+      Fin {r=Cons (p, l)}
     in
-    m.iolist s rnil rcons
+    trampoline @@ m.iolist s rnil rcons
 
  end
diff --git a/pretyping/constr_matching.ml b/pretyping/constr_matching.ml
index e511c92120..e2f809db79 100644
--- a/pretyping/constr_matching.ml
+++ b/pretyping/constr_matching.ml
@@ -529,14 +529,24 @@ let matches_head env sigma pat c =
     | _ -> c in
   matches env sigma pat head
 
+type 'a thunk =
+    | Fin of 'a
+    | TailCall of (unit -> 'a thunk)
+
+let rec trampoline r = match r with
+    | Fin a -> a
+    | TailCall f -> trampoline (f ())
+
+let tlift f () = trampoline (f ())
+
 (* Tells if it is an authorized occurrence and if the instance is closed *)
 let authorized_occ env sigma closed pat c mk_ctx =
   try
     let subst = matches_core_closed env sigma pat c in
     if closed && Id.Map.exists (fun _ c -> not (closed0 sigma c)) (snd subst)
-    then (fun next -> next ())
-    else (fun next -> mkresult subst (lazy (mk_ctx (mkMeta special_meta))) next)
-  with PatternMatchingFailure -> (fun next -> next ())
+    then (fun next -> TailCall next)
+    else (fun next -> Fin (mkresult subst (lazy (mk_ctx (mkMeta special_meta))) (tlift next)))
+  with PatternMatchingFailure -> (fun next -> TailCall next)
 
 let subargs env v = Array.map_to_list (fun c -> (env, c)) v
 
@@ -551,35 +561,35 @@ let sub_match ?(closed=true) env sigma pat c =
       | [c1] -> mk_ctx (mkCast (c1, k, c2))
       | _ -> assert false
       in
-      try_aux [env, c1] next_mk_ctx next
+      TailCall (fun () -> try_aux [env, c1] next_mk_ctx next)
   | Lambda (x,c1,c2) ->
       let next_mk_ctx = function
       | [c1; c2] -> mk_ctx (mkLambda (x, c1, c2))
       | _ -> assert false
       in
       let env' = EConstr.push_rel (LocalAssum (x,c1)) env in
-      try_aux [(env, c1); (env', c2)] next_mk_ctx next
+      TailCall (fun () -> try_aux [(env, c1); (env', c2)] next_mk_ctx next)
   | Prod (x,c1,c2) ->
       let next_mk_ctx = function
       | [c1; c2] -> mk_ctx (mkProd (x, c1, c2))
       | _ -> assert false
       in
       let env' = EConstr.push_rel (LocalAssum (x,c1)) env in
-      try_aux [(env, c1); (env', c2)] next_mk_ctx next
+      TailCall (fun () -> try_aux [(env, c1); (env', c2)] next_mk_ctx next)
   | LetIn (x,c1,t,c2) ->
       let next_mk_ctx = function
       | [c1; c2] -> mk_ctx (mkLetIn (x, c1, t, c2))
       | _ -> assert false
       in
       let env' = EConstr.push_rel (LocalDef (x,c1,t)) env in
-      try_aux [(env, c1); (env', c2)] next_mk_ctx next
+      TailCall (fun () -> try_aux [(env, c1); (env', c2)] next_mk_ctx next)
   | App (c1,lc) ->
      let lc1 = Array.sub lc 0 (Array.length lc - 1) in
      let app = mkApp (c1,lc1) in
      let mk_ctx = function
        | [app';c] -> mk_ctx (mkApp (app',[|c|]))
        | _ -> assert false in
-     try_aux [(env, app); (env, Array.last lc)] mk_ctx next
+     TailCall (fun () -> try_aux [(env, app); (env, Array.last lc)] mk_ctx next)
   | Case (ci,u,pms,hd0,iv,c1,lc0) ->
       let (mib, mip) = Inductive.lookup_mind_specif env ci.ci_ind in
       let (_, (hd,hdr), _, _, br) = expand_case env sigma (ci, u, pms, hd0, iv, c1, lc0) in
@@ -604,7 +614,7 @@ let sub_match ?(closed=true) env sigma pat c =
       | _ -> assert false
       in
       let sub = (env, c1) :: Array.fold_right (fun c accu -> (env, c) :: accu) pms (hd :: lc) in
-      try_aux sub next_mk_ctx next
+      TailCall (fun () -> try_aux sub next_mk_ctx next)
   | Fix (indx,(names,types,bodies as recdefs)) ->
     let nb_fix = Array.length types in
     let next_mk_ctx le =
@@ -612,7 +622,7 @@ let sub_match ?(closed=true) env sigma pat c =
       mk_ctx (mkFix (indx,(names, Array.of_list ntypes, Array.of_list nbodies))) in
     let env' = push_rec_types recdefs env in
     let sub = subargs env types @ subargs env' bodies in
-    try_aux sub next_mk_ctx next
+    TailCall (fun () -> try_aux sub next_mk_ctx next)
   | CoFix (i,(names,types,bodies as recdefs)) ->
     let nb_fix = Array.length types in
     let next_mk_ctx le =
@@ -620,11 +630,11 @@ let sub_match ?(closed=true) env sigma pat c =
       mk_ctx (mkCoFix (i,(names, Array.of_list ntypes, Array.of_list nbodies))) in
     let env' = push_rec_types recdefs env in
     let sub = subargs env types @ subargs env' bodies in
-    try_aux sub next_mk_ctx next
+    TailCall (fun () -> try_aux sub next_mk_ctx next)
   | Proj (p,_,c') ->
     begin match Retyping.expand_projection env sigma p c' [] with
-    | term -> aux env term mk_ctx next
-    | exception Retyping.RetypeError _ -> next ()
+    | term -> TailCall (fun () -> aux env term mk_ctx next)
+    | exception Retyping.RetypeError _ -> TailCall next
     end
   | Array(u, t, def, ty) ->
     let next_mk_ctx = function
@@ -632,9 +642,9 @@ let sub_match ?(closed=true) env sigma pat c =
     | _ -> assert false
     in
     let sub = (env,def) :: (env,ty) :: subargs env t in
-    try_aux sub next_mk_ctx next
+    TailCall (fun () -> try_aux sub next_mk_ctx next)
   | Construct _|Ind _|Evar _|Const _|Rel _|Meta _|Var _|Sort _|Int _|Float _|String _ ->
-    next ()
+    TailCall next
   in
   here next
 
@@ -642,16 +652,16 @@ let sub_match ?(closed=true) env sigma pat c =
   and try_aux lc mk_ctx next =
     let rec try_sub_match_rec lacc lc =
       match lc with
-      | [] -> next ()
+      | [] -> TailCall next
       | (env, c) :: tl ->
         let mk_ctx ce = mk_ctx (List.rev_append lacc (ce :: List.map snd tl)) in
         let next () = try_sub_match_rec (c :: lacc) tl in
-        aux env c mk_ctx next
+        TailCall (fun () -> aux env c mk_ctx next)
     in
-    try_sub_match_rec [] lc in
-  let lempty () = IStream.Nil in
+    TailCall (fun () -> try_sub_match_rec [] lc) in
+  let lempty () = Fin (IStream.Nil) in
   let result () = aux env c (fun x -> x) lempty in
-  IStream.thunk result
+  IStream.thunk (tlift result)
 
 let match_subterm env sigma pat c = sub_match env sigma pat c
 
